File: program/find-the-inverse-of-a/FindTheInverseOfA.java
import java.util.*;

public class FindTheInverseOfA {

    public static double[][] inverse(double[][] matrix) {
        int n = matrix.length;
        double[][] inverse = new double[n][n];

        // create the identity matrix
        for (int i = 0; i < n; i++) {
            inverse[i][i] = 1;
        }

        // augment the matrix with the identity matrix
        double[][] augmentedMatrix = new double[n][2 * n];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                augmentedMatrix[i][j] = matrix[i][j];
            }
            for (int j = n; j < 2 * n; j++) {
                augmentedMatrix[i][j] = inverse[i][j - n];
            }
        }

        // perform Gaussian elimination
        for (int i = 0; i < n; i++) {
            double pivot = augmentedMatrix[i][i];
            for (int j = i; j < 2 * n; j++) {
                augmentedMatrix[i][j] /= pivot;
            }
            for (int k = 0; k < n; k++) {
                if (k != i) {
                    double factor = augmentedMatrix[k][i];
                    for (int j = i; j < 2 * n; j++) {
                        augmentedMatrix[k][j] -= factor * augmentedMatrix[i][j];
                    }
                }
            }
        }

        // extract the inverse from the augmented matrix
        for (int i = 0; i < n; i++) {
            for (int j = n; j < 2 * n; j++) {
                inverse[i][j - n] = augmentedMatrix[i][j];
            }
        }

        return inverse;
    }

    public static void main(String[] args) {
        double[][] matrix = {{1,2,3},{4,5,6},{7,8,9}};
        double[][] inverse = inverse(matrix);

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                System.out.print(inverse[i][j] + " ");
            }
            System.out.println();
        }
    }
}

File: program/program/find-the-inverse-of-a/find_the_inverse_of_a.pl
use strict;
use warnings;

sub matrix_inverse {
    my ($matrix) = @_;

    my $size = scalar(@$matrix);

    # Check if the matrix is square
    die "Matrix must be square for inversion" if $size != scalar(@{$matrix->[0]});

    # Augment the matrix with an identity matrix
    my @augmented_matrix;
    for my $i (0..$size-1) {
        for my $j (0..$size-1) {
            $augmented_matrix[$i][$j] = $matrix->[$i][$j];
            $augmented_matrix[$i][$j+$size] = ($i == $j) ? 1 : 0;
        }
    }

    # Perform Gauss-Jordan elimination
    for my $i (0..$size-1) {
        my $pivot = $augmented_matrix[$i][$i];

        die "Matrix is singular, cannot find inverse" if $pivot == 0;

        for my $j (0..2*$size-1) {
            $augmented_matrix[$i][$j] /= $pivot;
        }

        for my $k (0..$size-1) {
            next if $k == $i;
            my $factor = $augmented_matrix[$k][$i];
            for my $j (0..2*$size-1) {
                $augmented_matrix[$k][$j] -= $factor * $augmented_matrix[$i][$j];
            }
        }
    }

    # Extract the inverse matrix from the augmented matrix
    my @inverse_matrix;
    for my $i (0..$size-1) {
        for my $j ($size..2*$size-1) {
            push @{$inverse_matrix[$i]}, $augmented_matrix[$i][$j];
        }
    }

    return @inverse_matrix;
}

# Example usage
my @matrix = (
    [4, 2, 7],
    [1, 5, 3],
    [8, 6, 9]
);

my @inverse_matrix = matrix_inverse(\@matrix);

# Print the inverse matrix
for my $i (0..$#inverse_matrix) {
    for my $j (0..$#{$inverse_matrix[$i]}) {
        printf "%.16f ", $inverse_matrix[$i][$j];
    }
    print "\n";
}

